pmd-core/src/main/java/net/sourceforge/pmd/benchmark/TimeTracker.java

```diff
@@ -0,0 +1,296 @@
+        }
+        
+        finishThread(); // finish the main thread
+        trackTime = false;
+        
+        // Fix UNACCOUNTED metric (total time is meaningless as is call count)
+        final TimedResult unaccountedResult = ACCUMULATED_RESULTS.get(
+                new TimedOperationKey(TimedOperationCategory.UNACCOUNTED, null));
+        unaccountedResult.totalTime.set(unaccountedResult.selfTime.get());
+        unaccountedResult.callCount.set(0);
+        
+        return new TimingReport(System.currentTimeMillis() - wallClockStartMillis, ACCUMULATED_RESULTS);
+    }
+    
+    /**
+     * Initialize a thread, starting to track it's own time.
+     */
+    public static TimedOperation initThread() {
+        if (!trackTime) {
+            return NOOP_TIMED_OPERATION;
+        }
+        
+        return startOperation(TimedOperationCategory.UNACCOUNTED);
+    }
+    
+    /**
+     * Finishes tracking a thread.
+     */
+    public static void finishThread() {
+        if (!trackTime) {
+            return;
+        }
+        
+        finishOperation(0);
+        
+        // if using a mono-thread, we may not be empty...
+        if (TIMER_ENTRIES.get().isEmpty()) {
+            TIMER_ENTRIES.remove();
+        }
+    }
+    
+    /**
+     * Starts tracking an operation.
+     * @param category The category under which to track the operation.
+     */
+    public static TimedOperation startOperation(final TimedOperationCategory category) {
+        return startOperation(category, null);
+    }
+    
+    /**
+     * Starts tracking an operation.
+     * @param category The category under which to track the operation.
+     * @param label A label to be added to the category. Allows to differentiate measures within a single category.
+     */
+    public static TimedOperation startOperation(final TimedOperationCategory category, final String label) {
+        if (!trackTime) {
+            return NOOP_TIMED_OPERATION;
+        }
+        
+        TIMER_ENTRIES.get().add(new TimerEntry(category, label));
+        return new TimedOperationImpl();
+    }
+    
+    /**
+     * Finishes tracking an operation.
+     * @param extraDataCounter An optional additional data counter to track along the measurements.
+     *                         Users are free to track any extra value they want (ie: number of analyzed nodes,
+     *                         iterations in a loop, etc.)
+     */
+    /* default */ static void finishOperation(final long extraDataCounter) {
+        if (!trackTime) {
+            return;
+        }
+
+        final Queue<TimerEntry> queue = TIMER_ENTRIES.get();
+        final TimerEntry timerEntry = queue.remove();
+
+        // Compute if absent
+        TimedResult result = ACCUMULATED_RESULTS.get(timerEntry.operation);
+        if (result == null) {
+            ACCUMULATED_RESULTS.putIfAbsent(timerEntry.operation, new TimedResult());
+            result = ACCUMULATED_RESULTS.get(timerEntry.operation);
```

Comment:
Is it possible to assign from the `putIfAbsent` call with a `get`?

Reply:
`putIfAbsent` returns null if no value is currently stored (in which case we would want to keep the new `TimedResult`) but returns the stored `TimedResult` if present (ie: a race condition let another thread put it between the current thread doing the get + null-check and getting to the `putIfAbsent`. So without a second variable and a null-check for the result of `putIfAbsent` (or just locking the whole Map) we can't avoid the `get`. All in all, I believe the `get` to be more concise and cleaner

Reply:
I missed the `null` if not present part, only checked return type on API doc.  Too complex, as you said.